x/participation/keeper/msg_participate.go

```diff
@@ -12,18 +12,31 @@ import (
 
 func (k msgServer) Participate(goCtx context.Context, msg *types.MsgParticipate) (*types.MsgParticipateResponse, error) {
 	ctx := sdk.UnwrapSDKContext(goCtx)
+	blockTime := ctx.BlockTime()
 
 	availableAlloc, err := k.GetAvailableAllocations(ctx, msg.Participant)
 	if err != nil {
 		return nil, err
 	}
 
 	// check if auction exists
-	_, found := k.fundraisingKeeper.GetAuction(ctx, msg.AuctionID)
+	auction, found := k.fundraisingKeeper.GetAuction(ctx, msg.AuctionID)
 	if !found {
 		return nil, sdkerrors.Wrapf(types.ErrAuctionNotFound, "auction %d not found", msg.AuctionID)
 	}
 
+	// check if auction is already started
+	if auction.IsAuctionStarted(blockTime) {
+		return nil, sdkerrors.Wrapf(types.ErrParticipationNotAllowed, "auction %d is already started", msg.AuctionID)
+	}
+
+	// check if auction allows participation at this time
+	registrationPeriod := k.RegistrationPeriod(ctx)
+	// as commented in `Time.Sub()`: To compute t-d for a duration d, use t.Add(-d).
+	if !blockTime.After(auction.GetStartTime().Add(-registrationPeriod)) {
```

Comment:
I may think `StartTime` is arbitrarily chosen by the auction creator, we need to ensure there are no edge case where `Add(-registrationPeriod)` is sensitive

Example:  `StartTime` == Unix 0 time (January 1th 1970), does it give a valid time if subtracted by the registration period

We can eventually discuss if `StartTime` should be > current time for auction creation

Reply:
the check `auction.IsAuctionStarted(blockTime)` done before would make this practically impossible, since `blockTime` should be a valid Unix time much greater than 1970 (which basically means we ensure that `startTime` is also greater than that). 

Moreover, `time.Duration` can be negative (see `(t Time) Sub()` and `minDuration`) so I think we are safe here.

We have two ways of dealing with this:

1. truncate negative values so that they are set to 0 (meaning that blockTime will always be after that)
2. return a critical error

I argue for the first. since the registration period is a param, we assume it will be correctly set by the community. And if the value is too big, that means registration is always possible.

thoughts?

Reply:
Yes `1` sounds good to me
This is a edge case that can't happen in practice but it is reachable in theory by design based on the possible values for the param we should still think about them
We can eventually think back about setting a large max range for the registration period like one years

Reply:
Let me know if how I addressed it seems fair.
I am not sure what would be the best approach.

Perhaps adding back a max value could actually be the safest solution

Reply:
The way it is addressed is fine IMO
This would happen in practice but since `block.Time()` and `StartTime` represent any timestamp we must handle edge cases allowed by design